packages/multichain/src/adapters/caip-permission-adapter-middleware.ts

```diff
@@ -0,0 +1,77 @@
+import type { NetworkConfiguration } from '@metamask/network-controller';
+import type { Caveat } from '@metamask/permission-controller';
+import { providerErrors } from '@metamask/rpc-errors';
+import type { JsonRpcRequest } from '@metamask/utils';
+
+import type { Caip25CaveatValue } from '../caip25Permission';
+import {
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+} from '../caip25Permission';
+import { mergeScopes } from '../scope/transform';
+import type { ScopeString } from '../scope/types';
+
+/**
+ * Middleware to handle CAIP-25 permission requests.
+ *
+ * @param request - The request object.
+ * @param _response - The response object.
+ * @param next - The next middleware function.
+ * @param end - The end function.
+ * @param hooks - The hooks object.
+ * @param hooks.getCaveat - Function to retrieve a caveat.
+ * @param hooks.getNetworkConfigurationByNetworkClientId - Function to retrieve a network configuration.
+ */
+export async function caipPermissionAdapterMiddleware(
+  request: JsonRpcRequest & {
+    networkClientId: string;
```

Comment:
total nit but we could pull in the `NetworkClientId` type here from `NetworkController` now

Reply:
https://github.com/MetaMask/core/pull/4784/commits/df5167ebef9fffb60b8caf57544346554edc7ae5